### ISSUE TITLE
Out of memory just with Require's

### ISSUE THREAD

**Issue by fblanqui** [coq/coq#18891]
### Description of the problem

Hi. I have an example file consisting of just 54 Require's that Coq cannot check even with 64 Gb RAM: the process is stopped for lack of memory.

In https://github.com/coq/coq/issues/17996, @charguer mentioned that: The changelog for Coq 8.4 reads: "Opaque proofs are now loaded lazily by default. This allows to be almost as fast as -dont-load-proofs".

But it doesn't seem to be the case as shown by my example file.

Would it be possible to add back the option `-dont-load-proofs` so that proofs are not saved in vo files?

It would be very helpful in my application, the translation of HOL-Light libraries to Coq with https://github.com/deducteam/hol2dk, which generates very big proofs.

### Version of Coq where this bug occurs

8.19.1

**Comment by charguer** [comment]
If you can't load all the proofs, I speculate that you have little hope of ultimately checking universe consistency.
In that case, the best you can achieve is a per-file verification in Coq. 
In this case, you could generate two versions of each file : the version with statements and proofs, and the version with only statements as axioms. 
When a file requires another file, use a "require import" on the version of the required file that contains only axioms.
This is a bullet-proof technique for not loading any proof :)

**Comment by fblanqui** [comment]
Yes, this is what I was thinking to do, but this is sad to have to do this...

**Comment by ppedrot** [comment]
Do you have a reproducible example? There are many reasons why Require can go wrong, not only proof loading. Also, opaque proofs are indeed not loaded until they are used (and Require doesn't AFAIK) so the problem must be elsewhere.

**Comment by ejgallego** [comment]
@fblanqui do you have an example of such files around?

**Comment by fblanqui** [comment]
It is kind of difficult to extract a minimum set of files among the thousands I have (in total 113 Gb in 120567 files) but I will try. What is strange is that the required vo files are not so big:
```
328K	coq.vo
24K	theory_hol.vo
24K	make_upto_topology_types.vo
808K	make_upto_topology_terms.vo
32K	make_upto_topology_axioms.vo
2,0M	HOM_INDUCED_CHAIN_MAP.vo
28K	NOT_CLAUSES_WEAK.vo
20K	thm0.vo
28K	thm10578.vo
48K	thm10597.vo
28K	thm16474.vo
44K	thm16506.vo
24K	thm16507.vo
60K	thm17045.vo
24K	thm1809.vo
24K	thm1810.vo
24K	thm1815.vo
24K	thm1816.vo
24K	thm1821.vo
48K	thm18394.vo
24K	thm1842.vo
24K	thm1845.vo
24K	thm1862.vo
24K	thm1863.vo
40K	thm19792.vo
88K	thm20230.vo
24K	thm20904.vo
48K	thm21007.vo
32K	thm21021.vo
24K	thm21114.vo
48K	thm21182.vo
44K	thm4211.vo
2,4M	thm58788437.vo
24K	thm58788438.vo
876K	thm58788452.vo
24K	thm58788453.vo
872K	thm58788467.vo
24K	thm58788468.vo
576K	thm58788482.vo
24K	thm58788483.vo
112K	thm58788560.vo
24K	thm58788561.vo
1,7M	thm58788602.vo
24K	thm58788603.vo
1,7M	thm58788627.vo
124K	thm58789095.vo
28K	thm58789108.vo
28K	thm58789113.vo
24K	thm58789114.vo
1,7M	thm58789120.vo
24K	thm7.vo
28K	thm82.vo
32K	thm58793604_part_1_type_abbrevs.vo
168K	thm58793604_part_1_term_abbrevs_part_1.vo
15M	total
```
Why requiring those files take more than 15M memory?

**Comment by fblanqui** [comment]
Is there any recursive behavior hidden? What does a Require? Does it load over things than what is the vo files? Note that I use no Require Export but just Require Import.

**Comment by JasonGross** [comment]
> Why requiring those files take more than 15M memory?

Coq loads vo files recursively.  If all of those .vo files are self-contained, then it's strange, but if these 15 files depend on your 113 gb of vo files, then it is not so strange that loading them all night take 113 gb of RAM

**Comment by fblanqui** [comment]
Does any one know how to extract all the dependencies (transitive closure) of a given file from a Makefile?

**Comment by JasonGross** [comment]
Try `make -B --dry-run foo.vo`.  -B says "force rebuild everything", --dry-run says "but don't actually change anything"

**Comment by fblanqui** [comment]
Thanks! The transitive closure has 19603 files. The total size of vo files is 21 Gb. This is <<64 Gb.

**Comment by fblanqui** [comment]
Difficult to share this. It takes many hours to check (>> 12 hours).

**Comment by fblanqui** [comment]
Does it make a difference in Require to use
```
Lemma l:A. Proof. exact t. Qed.
```
or
```
Definition l := t. Opaque l.
```
? If, in the second case, the definition of l is always loaded in memory while, in the first case, it is loaded only when necessary, this may explain the problem.

**Comment by JasonGross** [comment]
> Does it make a difference in Require to use
> 
> ```
> Lemma l:A. Proof. exact t. Qed.
> ```
> 
> or
> 
> ```
> Definition l := t. Opaque l.
> ```
> 
> ? If, in the second case, the definition of l is always loaded in memory while, in the first case, it is loaded only when necessary, this may explain the problem.

Yes, this is the case. `Opaque` does not really make things opaque, it just tells Coq to unfold them as late as possible and to make tactics error if you explicitly try to unfold them.

**Comment by ejgallego** [comment]
Yes, `Opaque` is reversible so indeed that's why we were asking about a concrete code sample.
